File: millipede-test/src/test/java/millipede/ntrip/Ntrip1Client.java
package millipede.ntrip;

import java.io.BufferedReader;
import java.io.IOException;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;

public class Ntrip1Client {
    private final String mount;
    private final String user;
    private final String password;
    private final String eol;

    private final Socket socket;

    public Ntrip1Client(String mount, String login, String password, String eol) {
        this.mount = mount;
        this.user = login;
        this.password = password;
        this.eol = eol;

        socket=initializeSocket();
    }

    public void stop(){
        try {
            socket.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private Socket initializeSocket() {
        String request = """
                GET /$mount HTTP/1.0
                User-Agent: NTRIP LOADTEST/0.0.0
                Authorization: Basic $basic
                
                """.replace("$mount", mount)
                .replace("$basic", java.util.Base64.getEncoder().encodeToString((user + ":" + password).getBytes()))
                .replace("\n",eol);

        try {
            var sock = new Socket("localhost", 2101);
            var os = sock.getOutputStream();
            var in = sock.getInputStream();

            os.write(request.getBytes());

            BufferedReader br = new BufferedReader(new java.io.InputStreamReader(in));
            String s;
            while (!(s=br.readLine()).isEmpty()) {
                System.out.println(s);
            }
            return sock;
        }catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    public List<String> getLineReceived() throws IOException{
        // Wait 50ms is mandatory because of the default 40ms TCP ACK delay in all the OS (congestion control)
        try {
            Thread.sleep(50);
        }catch(InterruptedException e){
            Thread.currentThread().interrupt();
        }
        var in = socket.getInputStream();
        BufferedReader br = new BufferedReader(new java.io.InputStreamReader(in));

        var lineReceived=new ArrayList<String>();
        while (br.ready()) {
            lineReceived.add(br.readLine());
        }
        return lineReceived;
    }
}

File: caster-test/steps/utils/ntrip1Client.py
import select
import socket
import base64
import time

class Ntrip1Client:
    def __init__(self, mount, user, password, eol):
        self.mount = mount
        self.user = user
        self.password = password
        self.eol = eol
        self.socket = self.initialize_socket()

    def stop(self):
        try:
            self.socket.close()
        except Exception as e:
            print(e)

    def initialize_socket(self):
        request = (
            f"GET /{self.mount} HTTP/1.0\n"
            f"User-Agent: NTRIP LOADTEST/0.0.0\n"
            f"Authorization: Basic {base64.b64encode((self.user + ':' + self.password).encode()).decode()}\n"
            "\n"
        ).replace("\n", self.eol)

        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.connect(("localhost", 2101))
            sock.setblocking(False)
            os = sock.makefile('wb')
            os.write(request.encode())
            os.flush()

            in_stream = sock.makefile('rb')
            while True:
                ready = select.select([sock], [], [], 0.5)  # Wait for the socket to be ready for reading
                if ready[0]:
                    while True:
                        s = in_stream.readline().strip()
                        if not s:
                            break
                        print(s.decode())
                    return sock
                else:
                    print("Client not ready after 0.5s. Exiting")
                    return None
        except Exception as e:
            print(e)
            return None

    def get_line_received(self):
        try:
            in_stream = self.socket.makefile('rb')
            line_received = []
            while True:
                ready = select.select([self.socket], [], [], 0.5)  # Wait for the socket to be ready for reading
                if ready[0]:
                    while True:
                        line = in_stream.readline().strip()
                        if not line:
                            break
                        line_received.append(line.decode())
                else:
                    break
            return line_received
        except Exception as e:
            print(e)
            return []
